### PR TITLE
MODPATBLK-18 Use calendar to determine loan's overdue status

### PR DESCRIPTION
**Pull request by alexanderkurash**
Resolves [MODPATBLK-18](https://issues.folio.org/browse/MODPATBLK-18).

Uses calendar for overdue minutes calculation to match similar behavior implemented in mod-circulation with CIRC-548.

### CODE REVIEW

File: src/main/java/org/folio/domain/Condition.java [hunk 1]
```diff
@@ -1,66 +1,62 @@
...
-      .count() >= limit.getValue()
+    (ctx) -> ctx.getUserSummary().getOpenLoans().stream()
+      .filter(openLoan -> ctx.getOverdueMinutes().get(openLoan.getLoanId()) > 0)
+      .count() >= ctx.getLimit().getValue()
   ),
 
   MAX_NUMBER_OF_OVERDUE_RECALLS("e5b45031-a202-4abb-917b-e1df9346fe2c",
-    (summary, limit) -> summary.getOpenLoans().stream()
-      .filter(Condition::isLoanOverdue)
+    (ctx) -> ctx.getUserSummary().getOpenLoans().stream()
```
Comment (roman-barannyk) [inline src/main/java/org/folio/domain/Condition.java:29]: remove ( )

File: src/main/java/org/folio/domain/Condition.java [hunk 2]
```diff
@@ -1,66 +1,62 @@
...
 
   RECALL_OVERDUE_BY_MAX_NUMBER_OF_DAYS("08530ac4-07f2-48e6-9dda-a97bc2bf7053",
-    (summary, limit) -> summary.getOpenLoans().stream()
-      .filter(Condition::isLoanOverdue)
+    (ctx) -> ctx.getUserSummary().getOpenLoans().stream()
+      .filter(openLoan -> ctx.getOverdueMinutes().get(openLoan.getLoanId()) > 0)
       .filter(OpenLoan::getRecall)
-      .map(Condition::getLoanOverdueDays)
-      .anyMatch(days -> days >= limit.getValue())
+      .map(openLoan -> getLoanOverdueDays(ctx.getOverdueMinutes().get(openLoan.getLoanId())))
```
Comment (roman-barannyk) [inline src/main/java/org/folio/domain/Condition.java:39]: please move getLoanOverdueDays to OverduePeriodCalculatorService

File: src/main/java/org/folio/domain/Condition.java [hunk 3]
```diff
@@ -1,66 +1,62 @@
...
-      .compareTo(BigDecimal.valueOf(limit.getValue())) >= 0
+    (ctx) -> ctx.getUserSummary().getOutstandingFeeFineBalance()
+      .compareTo(BigDecimal.valueOf(ctx.getLimit().getValue())) >= 0
   );
 
   private final String id;
-  private final BiPredicate<UserSummary, PatronBlockLimit> limitEvaluator;
+  private final Predicate<LimitEvaluationContext> limitEvaluator;
   private static final Map<String, Condition> idIndex = new HashMap<>(Condition.values().length);
+  private static final Double NUMBER_OF_MINUTES_IN_ONE_DAY = 1440.0;
```
Comment (roman-barannyk) [inline src/main/java/org/folio/domain/Condition.java:51]: move this constant to OverduePeriodCalculatorService

File: src/main/java/org/folio/rest/client/CalendarClient.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  private static final String PERIODS_QUERY_PARAMS = "servicePointId=%s&startDate=%s&endDate=%s&includeClosedDays=%s";
+  private static final String OPENING_PERIODS = "openingPeriods";
+  private static final String DATE_KEY = "date";
+  private static final String ALL_DAY_KEY = "allDay";
+  private static final String OPEN_KEY = "open";
+  private static final String OPENING_HOUR_KEY = "openingHour";
+  private static final String OPENING_DAY_KEY = "openingDay";
+  private static final String DATE_TIME_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+  private static final String DATE_PART_FORMAT = "yyyy-MM-dd";
+  private static final DateTimeFormatter DATE_TIME_FORMATTER =
```
Comment (roman-barannyk) [inline src/main/java/org/folio/rest/client/CalendarClient.java:48]: Do we need all these constants?

File: src/main/java/org/folio/rest/client/ConfigurationsClient.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    String query = cqlAnd(cqlExactMatch("module", "ORG"),
+      cqlExactMatch("configName", "localeSettings"));
+
+    try {
+      query = URLEncoder.encode(query, StandardCharsets.UTF_8.name());
+    }
+    catch (UnsupportedEncodingException e) {
+      String errorMessage = "Failed to encode query: " + query;
+      log.error(errorMessage);
+      return failedFuture(new EntityNotFoundException(errorMessage));
```
Comment (roman-barannyk) [inline src/main/java/org/folio/rest/client/ConfigurationsClient.java:53]: throw another type of exception

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+  public OverduePeriodCalculatorService(Vertx vertx, Map<String, String> okapiHeaders) {
+    this.calendarClient = new CalendarClient(vertx, okapiHeaders);
+    this.circulationStorageClient = new CirculationStorageClient(vertx, okapiHeaders);
+    this.feesFinesClient = new FeesFinesClient(vertx, okapiHeaders);
+    this.inventoryStorageClient = new InventoryStorageClient(vertx, okapiHeaders);
+  }
+
+  public Future<Integer> getMinutes(Loan loan, DateTime systemTime) {
+    if (loan.getOverdueFinePolicyId() == null) {
+      log.error("Loan ID {} - overdue fine policy is missing", loan.getId());
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/OverduePeriodCalculatorService.java:55]: return?

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+  private Future<Integer> minutesOverdueExcludingClosedPeriods(Loan loan, DateTime returnDate) {
+    DateTime dueDate = new DateTime(loan.getDueDate());
+    return getItemLocationPrimaryServicePoint(loan)
+      .compose(primaryServicePointId -> calendarClient
+        .fetchOpeningDaysBetweenDates(primaryServicePointId.toString(), dueDate, returnDate, false)
+        .map(openingDays -> getOpeningDaysDurationMinutes(openingDays, dueDate.toLocalDateTime(),
+          returnDate.toLocalDateTime())));
+  }
+
+  Integer getOpeningDaysDurationMinutes(
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/OverduePeriodCalculatorService.java:109]: private?

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+  private Future<Integer> minutesOverdueIncludingClosedPeriods(Loan loan, DateTime systemTime) {
+    int overdueMinutes = minutesBetween(new DateTime(loan.getDueDate()), systemTime).getMinutes();
+    return succeededFuture(overdueMinutes);
+  }
+
+  private Future<Integer> minutesOverdueExcludingClosedPeriods(Loan loan, DateTime returnDate) {
+    DateTime dueDate = new DateTime(loan.getDueDate());
+    return getItemLocationPrimaryServicePoint(loan)
+      .compose(primaryServicePointId -> calendarClient
+        .fetchOpeningDaysBetweenDates(primaryServicePointId.toString(), dueDate, returnDate, false)
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/OverduePeriodCalculatorService.java:104]: check NPE please

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+
+    return ZERO_MINUTES;
+  }
+
+  private int calculateDiffInMinutes(LocalDateTime start, LocalDateTime end) {
+    org.joda.time.Period period = new org.joda.time.Period(start, end);
+    return period.getHours() * MINUTES_PER_HOUR + period.getMinutes();
+  }
+
+  Integer adjustOverdueWithGracePeriod(CalculationContext context, int overdueMinutes) {
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/OverduePeriodCalculatorService.java:162]: private?

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+  }
+
+  private int getGracePeriodMinutes(CalculationContext context) {
+    return Optional.ofNullable(context)
+      .map(CalculationContext::getLoanPolicy)
+      .map(LoanPolicy::getLoansPolicy)
+      .map(LoansPolicy::getGracePeriod)
+      .map(gp -> Period.from(gp.getDuration(), gp.getIntervalId().value()))
+      .map(Period::toMinutes)
+      .orElse(0);
```
Comment (roman-barannyk) [inline src/main/java/org/folio/service/OverduePeriodCalculatorService.java:192]: you can use constant ZERO_MINUTES

File: src/main/java/org/folio/domain/Condition.java (hunk 1 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/domain/Condition.java]: Done

File: src/main/java/org/folio/domain/Condition.java (hunk 2 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/domain/Condition.java:41]: Done

File: src/main/java/org/folio/domain/Condition.java (hunk 3 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/domain/Condition.java]: Done

File: src/main/java/org/folio/rest/client/CalendarClient.java (hunk 4 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/rest/client/CalendarClient.java]: Removed

File: src/main/java/org/folio/rest/client/ConfigurationsClient.java (hunk 5 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/rest/client/ConfigurationsClient.java]: Done

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java (hunk 6 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/service/OverduePeriodCalculatorService.java]: Done

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java (hunk 7 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/service/OverduePeriodCalculatorService.java]: Done

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java (hunk 8 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/service/OverduePeriodCalculatorService.java:113]: Changed this piece - service point ID is passed as a method parameter now.

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java (hunk 9 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/service/OverduePeriodCalculatorService.java]: Done

File: src/main/java/org/folio/service/OverduePeriodCalculatorService.java (hunk 10 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/service/OverduePeriodCalculatorService.java]: Done

**Pull request merged**